File: packages/ios-im-ui/Sources/FlareIMUI/Components/ComposerParts.swift
import SwiftUI

/// Hold-to-talk voice button — a composable composer part. Press and hold to
/// record, release to send, slide up past the threshold to cancel. The host
/// owns the actual recording via the callbacks; all labels are host-provided.
public struct FlareVoiceHoldButton: View {
    private let label: String?
    private let recordingLabel: String?
    private let cancelLabel: String?
    private let cancelThreshold: CGFloat
    private let onStart: (() -> Void)?
    private let onEnd: (() -> Void)?
    private let onCancel: (() -> Void)?
    @Environment(\.colorScheme) private var scheme
    @Environment(\.flareBrandTheme) private var flareBrandTheme
    @Environment(\.flareStrings) private var strings
    @State private var pressing = false
    @State private var willCancel = false

    public init(label: String? = nil,
                recordingLabel: String? = nil,
                cancelLabel: String? = nil,
                cancelThreshold: CGFloat = 80,
                onStart: (() -> Void)? = nil, onEnd: (() -> Void)? = nil,
                onCancel: (() -> Void)? = nil) {
        self.label = label; self.recordingLabel = recordingLabel
        self.cancelLabel = cancelLabel; self.cancelThreshold = cancelThreshold
        self.onStart = onStart; self.onEnd = onEnd; self.onCancel = onCancel
    }

    public var body: some View {
        let colors = FlareColors.of(scheme, brand: flareBrandTheme)
        let bg = !pressing ? colors.bgSecondary : (willCancel ? colors.error : colors.primary)
        let fg = !pressing ? colors.textSecondary : Color.white
        Text(pressing ? (willCancel ? cancelLabel ?? strings.releaseToCancel : recordingLabel ?? strings.voiceHoldButtonRecording) : label ?? strings.voiceHoldButtonLabel)
            .font(.system(size: FlareSizes.fontSizeLg, weight: .medium))
            .foregroundColor(fg)
            .frame(maxWidth: .infinity, minHeight: 40)
            .background(RoundedRectangle(cornerRadius: FlareSizes.radiusXl).fill(bg))
            .gesture(
                DragGesture(minimumDistance: 0)
                    .onChanged { v in
                        if !pressing { pressing = true; willCancel = false; onStart?() }
                        let cancel = v.translation.height < -cancelThreshold
                        // Sliding past the line changes what letting go means, so it is felt as well as seen.
                        if flareHapticCrossed(was: willCancel, now: cancel) { flareHapticTick() }
                        if cancel != willCancel { willCancel = cancel }
                    }
                    .onEnded { _ in
                        let cancel = willCancel
                        pressing = false; willCancel = false
                        cancel ? onCancel?() : onEnd?()
                    }
            )
    }
}

/// One attachment/action tile in ``FlareComposerActionPanel``.
public struct FlareComposerAction: Identifiable {
    public let id: String
    public let label: String
    /// Semantic kit icon name (``flareIconNames``) on the tile.
    public let icon: String
    public let group: String?
    public let order: Int?
    public let visible: Bool
    public let enabled: Bool
    public let badge: String?
    public let intent: String?
    public let accessibilityLabel: String?
    public let disabledReason: String?
    public init(
        id: String,
        label: String,
        icon: String,
        group: String? = nil,
        order: Int? = nil,
        visible: Bool = true,
        enabled: Bool = true,
        badge: String? = nil,
        intent: String? = nil,
        accessibilityLabel: String? = nil,
        disabledReason: String? = nil
    ) {
        self.id = id; self.label = label; self.icon = icon
        self.group = group; self.order = order; self.visible = visible; self.enabled = enabled
        self.badge = badge; self.intent = intent; self.accessibilityLabel = accessibilityLabel
        self.disabledReason = disabledReason
    }
}

public struct FlareComposerCapabilities: Sendable, Equatable {
    public let availableActionIDs: Set<String>?
    public init(availableActionIDs: Set<String>? = nil) { self.availableActionIDs = availableActionIDs }
}

/// Defaults -> capabilities -> host list. A host list is a full replacement.
public func resolveComposerActions(
    defaults: [FlareComposerAction],
    capabilities: FlareComposerCapabilities = FlareComposerCapabilities(),
    actions: [FlareComposerAction]? = nil
) -> [FlareComposerAction] {
    var seen = Set<String>()
    return (actions ?? defaults).enumerated()
        .filter { _, action in
            action.visible && !action.id.isEmpty && seen.insert(action.id).inserted &&
                (capabilities.availableActionIDs?.contains(action.id) ?? true)
        }
        .sorted { left, right in
            let leftOrder = left.element.order ?? left.offset
            let rightOrder = right.element.order ?? right.offset
            return leftOrder == rightOrder ? left.offset < right.offset : leftOrder < rightOrder
        }
        .map(\.element)
}

/// The composer's bottom function area (下方功能区) — an inline grid of attachment
/// actions. Composable part; reveal it under the input. This is the one
/// attachment grid; the message long-press sheet is ``MessageActionSheetView``.
public struct FlareComposerActionPanel: View {
    private let actions: [FlareComposerAction]?
    private let capabilities: FlareComposerCapabilities
    private let columns: Int
    private let onAction: ((FlareComposerAction) -> Void)?
    @Environment(\.colorScheme) private var scheme
    @Environment(\.flareBrandTheme) private var flareBrandTheme
    @Environment(\.flareStrings) private var strings

    public init(actions: [FlareComposerAction]? = nil,
                capabilities: FlareComposerCapabilities = FlareComposerCapabilities(),
                columns: Int = 4, onAction: ((FlareComposerAction) -> Void)? = nil) {
        self.actions = actions; self.capabilities = capabilities
        self.columns = columns; self.onAction = onAction
    }

    /// Attachment tiles built from host-overridable copy. Pass a different
    /// ``FlareStrings`` (or the environment value) to relabel them.
    public static func actions(for strings: FlareStrings) -> [FlareComposerAction] {
        [
            .init(id: "image", label: strings.actionImage, icon: "image"),
            .init(id: "file", label: strings.actionFile, icon: "folder"),
            .init(id: "voice", label: strings.voiceHoldButtonLabel, icon: "mic"),
            .init(id: "location", label: strings.actionLocation, icon: "location"),
            .init(id: "contact", label: strings.actionCard, icon: "card"),
        ]
    }

    /// Default tiles with the kit's built-in copy. Prefer leaving `actions` unset so the
    /// view resolves them from the environment instead.
    public static let defaultActions: [FlareComposerAction] = actions(for: FlareStrings())

    public var body: some View {
        let colors = FlareColors.of(scheme, brand: flareBrandTheme)
        let actions = resolveComposerActions(
            defaults: Self.actions(for: strings),
            capabilities: capabilities,
            actions: self.actions
        )
        LazyVGrid(columns: Array(repeating: GridItem(.flexible()), count: columns),
                  spacing: FlareSizes.spacingLg) {
            ForEach(actions) { action in
                Button { onAction?(action) } label: {
                    VStack(spacing: FlareSizes.spacingXs) {
                        ZStack {
                            Color.clear.frame(width: 44, height: 44)
                            Image(systemName: flareIconSymbol(action.icon)).font(.system(size: FlareSizes.iconSizeMd))
                                .foregroundColor(colors.textPrimary)
                        }
                        Text(action.label).font(.system(size: FlareSizes.fontSizeXs))
                            .foregroundColor(colors.textSecondary)
                    }
                }
                .buttonStyle(.plain)
                .disabled(!action.enabled)
                .opacity(action.enabled ? 1 : FlareOpacity.disabled)
                .accessibilityLabel(action.accessibilityLabel ?? action.label)
                .accessibilityHint(action.enabled ? "" : (action.disabledReason ?? ""))
            }
        }
        .padding(FlareSizes.spacingLg)
        .frame(maxWidth: .infinity)
        .background(colors.bgPrimary)
    }
}

/// Send button (发送) — a composable composer part. Theme primary when active,
/// disabled otherwise; fires onSend only when active.
///
/// Same footprint as the send key inside ``ComposerView`` (36pt circle, 18pt
/// glyph) so a host composing its own bar gets an identical key.
/// Send — a paper plane, and nothing else.
///
/// It used to be a filled brand disc with a white arrow inside. Sending is the
/// same kind of act as every other key in the tool row — one tap, one outcome —
/// so it is drawn the same way, and only colour says which one sends: the brand
/// at rest against the row, faded while there is nothing to send. That is also
/// what `ComposerView`'s own send key does, and the two must not drift.
public struct FlareComposerSendButton: View {
    /// Shared with `ComposerView` so the two send keys can't drift.
    static let side: CGFloat = 44
    static let glyph: CGFloat = 20
    /// The plane at rest: present, but plainly not ready to be pressed.
    static let idleOpacity: Double = 0.38

    private let active: Bool
    private let onSend: (() -> Void)?
    @Environment(\.colorScheme) private var scheme
    @Environment(\.flareBrandTheme) private var flareBrandTheme
    @Environment(\.flareStrings) private var strings

    public init(active: Bool, onSend: (() -> Void)? = nil) {
        self.active = active
        self.onSend = onSend
    }

    public var body: some View {
        let colors = FlareColors.of(scheme, brand: flareBrandTheme)
        Button { if active { onSend?() } } label: {
            Image(systemName: "paperplane")
                .font(.system(size: Self.glyph))
                .frame(width: Self.side, height: Self.side)
                .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .foregroundColor(colors.primaryText.opacity(active ? 1 : Self.idleOpacity))
        .disabled(!active)
        .accessibilityLabel(Text(strings.send))
    }
}

/// Reply strip (回复条) — a composable composer part shown above the input when
/// replying. Left brand rail + sender / summary + cancel.
public struct FlareComposerReplyStrip: View {
    private let senderName: String
    private let summary: String
    private let label: String?
    private let onCancel: (() -> Void)?
    @Environment(\.colorScheme) private var scheme
    @Environment(\.flareBrandTheme) private var flareBrandTheme
    @Environment(\.flareStrings) private var strings

    public init(senderName: String, summary: String, label: String? = nil,
                onCancel: (() -> Void)? = nil) {
        self.senderName = senderName
        self.summary = summary
        self.label = label
        self.onCancel = onCancel
    }

    public var body: some View {
        let colors = FlareColors.of(scheme, brand: flareBrandTheme)
        HStack(spacing: 8) {
            VStack(alignment: .leading, spacing: 1) {
                Text("\(label ?? strings.composerReply) \(senderName)")
                    .font(.system(size: FlareSizes.fontSizeXs, weight: .semibold))
                    .foregroundColor(colors.primaryText)
                Text(summary)
                    .font(.system(size: FlareSizes.fontSizeSm))
                    .foregroundColor(colors.textSecondary)
                    .lineLimit(1)
            }
            Spacer(minLength: 0)
            Button { onCancel?() } label: {
                Image(systemName: flareIconSymbol("close")).font(.system(size: 18)).foregroundColor(colors.textTertiary)
                    .flareTouchTarget()
            }
            .buttonStyle(.plain)
            .flareCompactLayout(width: 18, height: 18)
            .accessibilityLabel(strings.cancelReply)
        }
        .padding(.horizontal, FlareSizes.spacingSm).padding(.vertical, FlareSizes.spacingXs)
        .background(
            RoundedRectangle(cornerRadius: FlareSizes.radiusMd).fill(colors.messageReplyBackground)
        )
        .overlay(
            HStack { Rectangle().fill(colors.messageReplyBorder).frame(width: 3); Spacer() }
                .clipShape(RoundedRectangle(cornerRadius: FlareSizes.radiusMd))
        )
    }
}

File: packages/android-im-ui/src/main/kotlin/com/flare/im/ui/ComposerParts.kt
package com.flare.im.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.gestures.awaitEachGesture
import androidx.compose.foundation.gestures.awaitFirstDown
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.Send
import androidx.compose.material.icons.rounded.Close
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * Hold-to-talk voice button (语音) — a composable composer part. Press and hold
 * to record, release to send; the host owns recording via the callbacks.
 */
@Suppress("NAME_SHADOWING")
@Composable
fun FlareVoiceHoldButton(
    label: String? = null,
    recordingLabel: String? = null,
    cancelLabel: String? = null,
    cancelThreshold: Float = 80f,
    onStart: (() -> Unit)? = null,
    onEnd: (() -> Unit)? = null,
    onCancel: (() -> Unit)? = null,
) {
    val strings = flareStrings()
    val label = label ?: strings.voiceHoldButtonLabel
    val recordingLabel = recordingLabel ?: strings.voiceHoldButtonRecording
    val cancelLabel = cancelLabel ?: strings.voiceHoldButtonCancel
    val colors = flareColors()
    var pressing by remember { mutableStateOf(false) }
    var willCancel by remember { mutableStateOf(false) }
    val haptics = LocalHapticFeedback.current
    val bg = if (!pressing) colors.bgSecondary else if (willCancel) colors.error else colors.primary
    Box(
        Modifier.fillMaxWidth().height(FlareSizes.componentComposerActionHeight)
            .clip(RoundedCornerShape(FlareSizes.radiusXl))
            .background(bg)
            .pointerInput(cancelThreshold) {
                val cancelPx = cancelThreshold * density
                awaitEachGesture {
                    val down = awaitFirstDown()
                    pressing = true
                    willCancel = false
                    onStart?.invoke()
                    var cancelled = false
                    while (true) {
                        val event = awaitPointerEvent()
                        val change = event.changes.firstOrNull { it.id == down.id } ?: break
                        // Sliding up past the threshold from the press origin arms cancel.
                        val c = (change.position.y - down.position.y) < -cancelPx
                        // Sliding past the line changes what letting go means, so it is felt as well as seen.
                        if (flareHapticCrossed(willCancel, c)) haptics.flareHapticTick()
                        if (c != willCancel) willCancel = c
                        cancelled = c
                        if (!change.pressed) break
                    }
                    pressing = false
                    willCancel = false
                    if (cancelled) onCancel?.invoke() else onEnd?.invoke()
                }
            },
        contentAlignment = Alignment.Center,
    ) {
        Text(
            if (!pressing) label else if (willCancel) cancelLabel else recordingLabel,
            color = if (pressing) Color.White else colors.textSecondary,
            fontSize = FlareSizes.fontSizeLg.value.sp,
            fontWeight = FontWeight.Medium,
        )
    }
}

/**
 * The composer's bottom function area (下方功能区) — an inline grid of attachment
 * actions. Composable part; wrap in `AnimatedVisibility` to expand under the
 * input, or use standalone. Shares [FlareComposerAction] with the action sheet.
 */
@Composable
fun FlareComposerActionPanel(
    actions: List<FlareComposerAction>? = null,
    capabilities: FlareComposerCapabilities = FlareComposerCapabilities(),
    columns: Int = 4,
    onAction: ((FlareComposerAction) -> Unit)? = null,
) {
    val colors = flareColors()
    val resolvedActions = resolveComposerActions(defaultComposerActions(), capabilities, actions)
    Column(Modifier.fillMaxWidth().background(colors.bgPrimary).padding(FlareSizes.spacingLg)) {
        resolvedActions.chunked(columns).forEach { row ->
            Row(
                Modifier.fillMaxWidth().padding(bottom = FlareSizes.spacingLg),
                horizontalArrangement = Arrangement.spacedBy(FlareSizes.spacingLg),
            ) {
                row.forEach { action ->
                    Column(
                        Modifier.weight(1f).alpha(if (action.enabled) 1f else FlareOpacity.disabled)
                            .clickable(enabled = action.enabled) { onAction?.invoke(action) },
                        horizontalAlignment = Alignment.CenterHorizontally,
                    ) {
                        Box(
                            Modifier.size(FlareSizes.componentComposerActionWidth).clip(RoundedCornerShape(FlareSizes.radiusLg))
                                .background(Color.Transparent),
                            contentAlignment = Alignment.Center,
                        ) { FlareIcon(name = action.icon, size = 20.dp, tint = colors.textPrimary, contentDescription = action.accessibilityLabel ?: action.label) }
                        Spacer(Modifier.size(FlareSizes.spacingXs))
                        Text(action.label, color = colors.textSecondary, fontSize = FlareSizes.fontSizeXs.value.sp)
                    }
                }
                repeat(columns - row.size) { Spacer(Modifier.weight(1f)) }
            }
        }
    }
}

/**
 * Send button (发送) — a composable composer part. Theme primary when active,
 * disabled otherwise; fires onSend only when active.
 */
/**
 * Send — a paper plane, and nothing else.
 *
 * It used to be a filled brand disc with a white glyph inside. Sending is the
 * same kind of act as every other key in the tool row — one tap, one outcome —
 * so it is drawn the same way, and only colour says which one sends: the brand
 * at rest against the row, faded while there is nothing to send. That is also
 * what [Composer]'s own send key does, and the two must not drift.
 */
@Composable
fun FlareComposerSendButton(active: Boolean, onSend: (() -> Unit)? = null) {
    val colors = flareColors()
    Box(
        Modifier.size(FlareSizes.componentComposerActionWidth).clickable(enabled = active) { onSend?.invoke() },
        contentAlignment = Alignment.Center,
    ) {
        Icon(
            Icons.AutoMirrored.Filled.Send, flareStrings().send, Modifier.size(20.dp),
            tint = colors.primaryText.copy(alpha = if (active) 1f else 0.38f),
        )
    }
}

/**
 * Reply strip (回复条) — a composable composer part shown above the input when
 * replying. Left brand rail + sender / summary + cancel.
 */
@Suppress("NAME_SHADOWING")
@Composable
fun FlareComposerReplyStrip(
    senderName: String,
    summary: String,
    label: String? = null,
    onCancel: (() -> Unit)? = null,
) {
    val strings = flareStrings()
    val label = label ?: strings.composerReplyStripLabel
    val colors = flareColors()
    Row(
        Modifier.fillMaxWidth().clip(RoundedCornerShape(FlareSizes.radiusMd))
            .background(colors.messageReplyBackground),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Box(Modifier.width(3.dp).height(36.dp).background(colors.messageReplyBorder))
        Column(
            Modifier.weight(1f).padding(horizontal = FlareSizes.spacingSm, vertical = FlareSizes.spacingXs),
        ) {
            Text(
                "$label $senderName", color = colors.primaryText,
                fontSize = FlareSizes.fontSizeXs.value.sp, fontWeight = FontWeight.SemiBold,
            )
            Text(summary, color = colors.textSecondary, fontSize = FlareSizes.fontSizeSm.value.sp, maxLines = 1)
        }
        Box(
            Modifier.padding(end = FlareSizes.spacingSm).size(18.dp).clickable { onCancel?.invoke() },
            contentAlignment = Alignment.Center,
        ) { Icon(Icons.Rounded.Close, flareStrings().cancelReply, Modifier.size(18.dp), tint = colors.textTertiary) }
    }
}
